module2/projects/hang-in-there-api/requirements.md

```diff
@@ -0,0 +1,579 @@
+* API will be compliant to the [JSON API spec](https://jsonapi.org/) and match our requirements below precisely
+* Controller actions should be limited to only the standard Rails actions and follow good RESTful convention.
+
+
+You will need to expose the following RESTful API endpoints for the following:
+
+<details><summary><h3>1. Fetch all Posters</h3></summary>
+These "index" endpoints should:
+
+* render a JSON representation of all records of the requested resource
+* always return an array of data, even if one or zero resources are found
+* follow this pattern: `GET /api/v1/posters`
+
+Example JSON response:
+
+```json
+{
+  "data": [
+    {
+      "id": "1",
+      "type": "poster",
+      "attributes": {
+        "name": "FAILURE",
+        "description": "Why bother trying? It's probably not worth it.",
+        "price": 68.00,
+        "year": 2019,
+        "vintage": true,
+        "img_url": "https://images.unsplash.com/photo-1620401537439-98e94c004b0d"
+      }
+    },
+    {
+      "id": "2",
+      "type": "poster",
+      "attributes": {
+        "name": "REGRET",
+        "description": "Hard work rarely pays off.",
+        "price": 89.00,
+        "year": 2018,
+        "vintage": true,
+        "img_url":  "https://plus.unsplash.com/premium_photo-1661293818249-fddbddf07a5d",
+      }
+    },
+    {
+      "id": "3",
+      "type": "poster",
+      "attributes": {
+        "name": "MEDIOCRITY",
+        "description": "Dreams are just that—dreams.",
+        "price": 127.00,
+        "year": 2021,
+        "vintage": false,
+        "img_url": "https://images.unsplash.com/photo-1551993005-75c4131b6bd8",
+      }
+    }
+  ]
+}
+```
+
+<hr/>
+</details>
+<details><summary><h3>2. Fetch a single record</h3></summary>
+
+This endpoint for Posters should:
+
+* render a JSON representation of the corresponding record, if found
+* follow this pattern: `GET /api/v1/poster/:id`
+
+Example JSON response:
+
+```json
+{
+  "data": {
+      "id": "1",
+      "type": "poster",
+      "attributes": {
+        "name": "FAILURE",
+        "description": "Why bother trying? It's probably not worth it.",
+        "price": 68.00,
+        "year": 2019,
+        "vintage": true,
+        "img_url": "https://images.unsplash.com/photo-1620401537439-98e94c004b0d"
+      }
+    }
+}
+```
+
+<hr/>
+</details>
+
+
+<details><summary><h3>3. Create a Poster</h3></summary>
+
+This endpoint should:
+
+* create a record and render a JSON representation of the new Poster record.
+* follow this pattern: `POST /api/v1/posters`
+* accept the following JSON body with only the following fields:
+
+```json
+{
+  "name": "DEFEAT",
+  "description": "It's too late to start now.",
+  "price": 35.00,
+  "year": 2023,
+  "vintage": false,
+  "img_url":  "https://unsplash.com/photos/brown-brick-building-with-red-car-parked-on-the-side-mMV6Y0ExyIk"
+}
+```
+* you should ignore any attributes sent by the user which are not allowed
```

Comment:
This is for my own curiosity.  Is ignoring unneeded attributes standard?  On the FE side, if I send an attribute that isn't supposed to be included, I often get a 422 error back.

Reply:
Yes and no. If our database doesn't have a spot for an attribute nothing would happen anyways, but also for security reasons, typically in rails we permit the sending of certain attributes. If an attribute exists in the database but we do NOT want to allow outside changes we simply don't permit it. For posters there isn't a great example of this but typically we don't allow the changing of the created/updated timestamps. Just because those attributes exist on the table, doesn't mean we want users to have the ability to make those changes. Then we can either ignore OR send back an error saying it's not allowed. Students will not have learned about errors yet so here we just want them to ignore.